Nevm

sidhujag opened the pull request.

Commit:
nevm wip

Commit:
add ModeNEVM

Commit:
remove eth66

Commit:
wip tests

Commit:
add seperate network for nevm network

Commit:
fix tests

Commit:
add test network stubbing

Commit:
fix test

Commit:
work uncle hash non-req for nevm

Commit:
gofmt'ed

Commit:
nevm shouldn't actively broadcast blocks only respond to queries

nevm should avoid broadcasting blocks because block validation is ultimately set by syscoin nodes for which blocks are valid through the mappings that are set in geth, instead when blocks are queried if they are mapped they can be responded to

Commit:
add nevm mapping logic

1) hook into verifyHeader for mapping check
2) move nevmindex into backend
3) db persistence/cache for nevm and sys mapping

Commit:
fix startup and verify header logic

Commit:
initial ZMQ wip

integrate github.com/go-zeromq/zmq4 full go zmq implementation and add the flags to geth to initiatilize zmq

Commit:
test zmq, working with custom pub/sub

Commit:
encode and output block for createBlockSub

Commit:
only write block to state if not nevm on miner

miner node should remain idempotent

Commit:
deserialize block connects

use btcd/wire from syscoin msgtxsyscoin.go to deserialize block connect messages from syscoin core on zmq layer

Commit:
add serailizations for block create/add

Commit:
fix NEVMBlockWire serialization + update btcd

Commit:
update fmt

Commit:
correctly refer to tx root instead of state root

Root() is state root and TxHash of header is tx root (the one we need)

Commit:
Update backend.go

Commit:
Update zmqpubsub.go

Commit:
only deser block and check its fields if there is data actually there

Commit:
use req/rep model

1) convert sysmapping key to string so its hashable ([]byte is not hashable)
2) delete cached mappings from headerchain when deleting mappings
3) fix addBlock logic
4) change deleteBlock to call WriteKnownBlock instaed of reorg since it wasn't writing to DB
5) update zmq4 to pull in rep bug fix
6) update btcd with new ser/der fixes in wire

Commit:
only add to chain if building on tip

if not building on tip we shouldn't add to chain.

Situation would happen if we are reindexing blocks and we are < maxtipage*2 so we won't hear about nevm block data (which means we only add to mappings). In this case we will end up skipping over blocks and when it comes to >= maxtipage*2 and we get nevm block data we will not be in sequential order. Instead we rely on Geth sync to fetch blocks and we only setup mappings (not insert into chain even if we have the block data)

Commit:
fix nevmcomms

send response in the case where nevmcomms is set to anything other than 0

Commit:
fix verifyHeader failure

1) recreate miner to avoid state issues, subsequent blocks were not allowed to be recreated once there was a failure, by resetting miner object it will allow to create subsequent blocks
2) only store sys mapping if its valid string > 0 length, don't need or want to store an empty string/hash

Commit:
enforce linear nevm mappings

ensure nevm mapping -> prev is = next mapping being proposed and on disconnect that the latest mapping = tip chain. It ties mappings to the canonical chain and also ensures mappings are linearly progressing fwd/reverse

Commit:
save db read using cache

use memory cache for latest nevm block

Commit:
add cache purge for consistency to NEVMLatestCache in SetHead

Commit:
skip networking startup by default and rely on when blocks get inserted to start it

this relies on syscoin core being the source of blocks until a block is inserted meaning likely initial block download is done, where peers can connect and source info from this node or sync with it in case syscoin core was syncing and only setting mappings in geth

Commit:
refactor nevm callbacks to take into account light client mode

1) move calls to headerchain and reference from blockchain/lightchain
2) create les version of callbacks and zmq subscriber
3) add NEVMPubEP endpoint string to config and call init inside of zmq constructor

Commit:
light chain disconnect logic

we need to reorg the light chain in disconnect to move the chain back

Commit:
use currentblock not header in full chain mode, headers not indicative of current statee after removing or adding blocks

tomatoskittles commented on cmd/geth/main.go:
```diff
@@ -417,10 +427,13 @@ func startNode(ctx *cli.Context, stack *node.Node, backend ethapi.Backend) {
-		threads := ctx.GlobalInt(utils.MinerThreadsFlag.Name)
-		if err := ethBackend.StartMining(threads); err != nil {
-			utils.Fatalf("Failed to start mining: %v", err)
+		// SYSCOIN
```
I'd add to the comment something about using a fake PoW with those flags, so we intentionally don't want to start mining. I know it sounds obvious, but just to be very clear.

Commit:
add comment around avoiding to start mining

Commit:
change nevmpub from miner to regular flag

Commit:
introduce new opcode SYSBLOCKHASH

SYSBLOCKHASH is responsible for taking in a block number (NEVM) (last 50000) blocks inclusive and giving back the Syscoin bytes[] representing the block hash. It is useful to know inside of solidity that an NEVM block is associated with the canonical chain of a Syscoin block so trustless proofs can be created for the bridge SYS->NEVM

Commit:
remove syshash check, not needed

Commit:
Update wizard_genesis.go

Commit:
fix miner related to sealing

1) disable presealing which creates a presealed block and noempty which removes empty block skipping
2) update max interval for commiting to 150 sec (2.5 min blocks)
3) add continious check against nevm parent hash since blocks were able to be created not continious to the actual nevm chain

Commit:
add tenenbaum allocations

Commit:
fix parenthash checks

check for parent hash only during mining so prevent the addblock logic from preventing connection later which invalidates block and causes miner to get stuck. The currentHash may be based on block that is old or genesis (during initial sync) and so we cannot assume current hash will be same as blockconnect.parenthash as only recent blocks will be connected by syscoin until geth syncs up the main chain through geth sync.

Commit:
Update bootnodes.go

Commit:
wip puppeth updates

Commit:
fix pupeth faucet to work with tanenbaum

Commit:
update link to syscoind in pupeth

Commit:
use focal instead of bionic

Commit:
remove gzip untarring

Commit:
remove escaping chars in ethstats

Commit:
hardcode datadir

Commit:
add datadir config in docker file

Commit:
fix lightFLag

Commit:
remove datadir

Commit:
add zmq nevm pub info when executing syscoind

Commit:
fix pupeth timing and remove path extension

Commit:
update syscoind link

Commit:
use regtest for now

Commit:
rework pupeth geth/syscoin builds using alpine

Commit:
fix docker images building core with alpine

Commit:
disable wallet when running syscoind

Commit:
Update module_explorer.go

Commit:
download descriptor to datadir prior to run

Commit:
add executable and read priv. for sysgeth

Commit:
fix disablewallet switch

Commit:
mkdir before wget

Commit:
fix typo in wget

Commit:
add wget output

Commit:
mk regtest folder for now not testnet3

Commit:
add chmod to descriptor

Commit:
put descriptor in base datadir where it belongs

Commit:
remove creation of network specific folder

Commit:
update bootnode

Commit:
add local node to connect

Commit:
update ports for syscoind

Commit:
expose syscoind ports and do not hardcode port to 8369 on startup

Commit:
export sys ports in composer file

Commit:
Update module_node.go

Commit:
update def peer

Commit:
expose sys ports

Commit:
remove hardcoded port

Commit:
add sys ports on docker deploy script for defaults

Commit:
update rewards

Commit:
Update bootnodes.go

Commit:
in nevm mode stop peering on start and start it after sync finished

IBD in core controls when sync should be finished, so in addblock when block is not nil that means sync is finished, we start peering then and networking stack. Prior when peering wasn't disabled we had peers connect and get banned because networking stack wasn't enabled so during long syncs the node wouldn't sync via another geth node

Commit:
def bootnode

Commit:
update network

Commit:
delay network/peering start to 5s since last block

delay to start networking until enough time has been passed to be sure that you are indeed synced after IBD.

Commit:
cleanup docker files

make syscoin-core alpine aware of geth alpine etc

Commit:
fix spaces

Commit:
remove expose of ports

Commit:
set timeLastBlock only if not inited to avoid extra locking

Commit:
use opt/add/.syscoin datadir home and rm old geth

Commit:
fix exitwhensynced cmd line

Commit:
fix for geth checkNode service

Commit:
move syscoind under mix calls to start explorer services

Commit:
Update module_explorer.go

Commit:
remove debug

Commit:
copy execs to syscoin datadir

Commit:
multisig bootnode

Commit:
use our own blockscout upstream

Commit:
fix data dir location for descriptor copy

Commit:
Update module_node.go

Commit:
fix docker images

Commit:
explicit datadir specification to syscoind

Commit:
add http api's for explorer to work

Commit:
Update module_explorer.go

Commit:
use env prior to calling explorer for network specific branding

Commit:
Update module_explorer.go

Commit:
Update module_explorer.go

Commit:
Update module_explorer.go

Commit:
Update module_explorer.go

Commit:
add supported chains and pricing chart config

Commit:
set ARGs cof COIN/BLOCK_TRANSFORMER for build script

Commit:
Update module_explorer.go

Commit:
add svg to files

Commit:
add host based http and ws url

Commit:
fix logoSVG file decl

Commit:
remove extra space

Commit:
flip supportChains logic

Commit:
remove spaces from supportedChains

Commit:
Update module_explorer.go

Commit:
update healthy block period to 3450 seconds

Commit:
fix wsURL

Commit:
Update module_explorer.go

Commit:
update svgFilePath

Commit:
put svg in static host section of vm where /images are found

Commit:
assume logo exists in blockscout

Commit:
add args in docker

Commit:
rebuild resources after css updates

Commit:
fix supportedChains args

Commit:
Update module_explorer.go

Commit:
sync network/peering logic with fullnode sync

Commit:
build explorer again after clean

Commit:
escape supportedChains string

Commit:
dedup COIN

Commit:
start server after geth synced

Commit:
run syscoind's at the beginning

Commit:
Update module_explorer.go

Commit:
Update module_explorer.go

Commit:
update max drift to 40s

Commit:
Update module_explorer.go

Commit:
SYS-49: Changed where the supported chains json is hard coded. This avoids the html template encoding the json, which makes it not work.

Commit:
SYS-49: Restoring the consistent whitespace.

Commit:
SYS-49: Another unintentional change reversed.

Commit:
SYS-49: Changing the template type to text, from html, removes the html escaping. This seems to be working.

Commit:
Merge pull request #2 from tomatoskittles/SYS-49

SYS-49: Changed where the supported chains json is hard coded. …

Commit:
SYS-54: Correcting both the location server.js is created in, and where it is executed from. There were mismatches between these, and also where the node modules were installed.

Commit:
SYS-54: Upon further testing, it was only the / in the execution which needed to be changed. The rest were all properly referencing the working directory.

Commit:
Merge pull request #3 from tomatoskittles/SYS-54

SYS-54: Correcting where server.js is executed from.

Commit:
update faucet to run syscoind

Commit:
Merge branch 'nevm' of https://github.com/sidhujag/go-ethereum into nevm

Commit:
fix faucet copy from alpine image

Commit:
add zmq pub sub url and datadir passthrough to faucet flag parsing

Commit:
faucet branding

Commit:
add genesis.json into cmd line args

Commit:
Update module_faucet.go

Commit:
Update website.go

Commit:
Revert "Update website.go"

This reverts commit 752ac590399c3bc1bd173ff177daec301024bfd1.

Commit:
Merge branch 'master' into nevm

Commit:
stop networking properly on node startup

holiman committed:
tests: update, enable legacy tests, remove vm tests (#23350)

* tests: update, enable legacy tests, remove vm tests

* tests: minor fixes

Commit:
add missing translation

jwasinger committed:
cmd, metrics: add support for influxdb-v2 (cherry-picking from italoa…

Commit:
cmd, core: remove calaveras testnet (#23366)

Commit:
cmd/utils: add --nousb to the list of deprecated flags (#23388)

Adds --nousb as a deprecated flag when someone runs the geth show-deprecated-flags command.

chenjia404 committed:
eth, internal/ethapi: make RPC block miner field show block sealer correctly (#23312)

Makes the RPC block return the POA sealer for clique blocks on the 'miner' field (was previously zeroes)

Commit:
SYS-57: The bootnode check looks for genesis.json so I added it back in, and also we need to specify the location of the ipc file to attach to sysgeth and get the enode address. I used a find because the location will change depending on the network specified.

chuwt committed:
eth/downloader: fix typo in comment (#23413)

Commit:
Merge pull request #5 from tomatoskittles/SYS-57

SYS-57: Fixing the puppeth bootnode check

Commit:
fix bigint equality check for chainid

faucet was not setting powmode here because the bigint equality checks will fail, convert to uint64 to do the check instead

Commit:
fix datadir override in faucet

Commit:
remove genesis and nodiscover to false

genesis is not needed as we have specific networks preloaded in sysgeth. We specify the network name anyway so no need for custom genesis.json. Also noticed we need to enable regular discovery for nodes to connect as I guess there is no v5 discovery servers running

Commit:
Update faucet.html

Commit:
Update website.go

Commit:
update dashboard branding

Commit:
Update module_dashboard.go

Commit:
fix end tag

Commit:
fix services stop on startmining

Commit:
fix whitespace

Commit:
use int64 of network id in dashboard

Commit:
Update module_dashboard.go

MariusVanDerWijden committed:
core: only check sendernoeoa in non fake mode (#23424)

fjl committed:
signer/core/apitypes: remove dependency on internal/ethapi (#23362)

ucwong committed:
go.mod: upgrade golang.org/x/sys for go1.17 support (#23406)

Commit:
add BLOCKSCOUT_PROTOCOL/BLOCKSCOUT_HOST

zsfelfoldi committed:
eth/gasprice: feeHistory improvements (#23422)

* eth/gasprice: cache feeHistory results

* eth/gasprice: changed feeHistory block count limitation

* eth/gasprice: do not use embedded struct in blockFees

* eth/gasprice: fee processing logic cleanup

* eth/gasprice: purge feeHistory cache at chain reorgs

holiman committed:
core/vm: faster code analysis (#23381)

* core/vm: more detailed benchmark for jumpdest analysis

* core/vm: make jd analysis benchmark alloc free

* core/vm: improve jumpdest analysis

* core/vm: improve worst-case

* core/vm: further improvements in analysis

* core/vm: improve jumpdest analysis >PUSH15

* core/vm: make jd analysis ref by value

* core/vm: fix misspell

* core/vm: improve set8 and set16 a bit

* core/vm: reduce amount of code

* core/vm: optimize byte copying

chuwt committed:
trie, les: add missing calls to Ticker.Stop (#23415)

MariusVanDerWijden committed:
p2p, p2p/enode: fix data races (#23434)

In p2p/dial.go, conn.flags was accessed without using sync/atomic.
This race is fixed by removing the access.

In p2p/enode/iter_test.go, a similar race is resolved by writing the field atomically.

Co-authored-by: Felix Lange <fjl@twurst.com>

otherview committed:
core/bloombits: avoid crash when storing errors of different type (#23437)

This fixes a rare crash which could occur when two different errors happened
in the same bloombits.MatcherSession.

ligi committed:
tests/testdata: make submodule shallow (#23412)

Before:

316M .git/modules/tests/

After:

171M .git/modules/tests/

gballet committed:
core/vm: fix typo in comment (#23450)

holiman committed:
core/types: add benchmarks for rlp encoding/decoding (#23190)


Co-authored-by: Felix Lange <fjl@twurst.com>
Co-authored-by: Sina Mahmoodi <1591639+s1na@users.noreply.github.com>

Commit:
node: add comment about --nousb being deprecated (#23439)


Co-authored-by: Felix Lange <fjl@twurst.com>

baptiste-b-pegasys committed:
tests/fuzzers: fix go vet warning about ReadByte (#23380)

holiman committed:
core: make txpool reject too sudden changes (#23095)

* core: make txpool reject too sudden changes

* core: add some metrics to txpool

bogatyy committed:
cmd/utils: fix typo in variable name (#23451)

fjl committed:
metrics: fix compilation for GOOS=js (#23449)

rjl493456442 committed:
core, light, tests, trie: add state metrics (#23433)

ucwong committed:
go.mod: update goleveldb (#23417)

Commit:
replace ethereum coin gecko ID which was hardcoded to the coingeck coin id for the project ("syscoin")

Commit:
Update module_explorer.go

rjl493456442 committed:
les: fix data races in tests (#23457)

fjl committed:
build: upgrade to golangci-lint v1.42.0 (#23461)

The new linter version is built with go 1.17 and thus includes the go vet
check for mismatched +build and go:build lines.

Fortunately, no new warnings are reported with this update.

fjl committed:
build, .travis.yml: upgrade to Go 1.17 (#23464)

holiman committed:
cmd/evm: add difficulty calculation to t8n tool (#23353)

This PR adds functionality to the evm t8n to calculate ethash difficulty. If the caller does not provide a currentDifficulty, but instead provides the parentTimestamp (well, semi-optional, will default to 0 if not given), and parentDifficulty, we can calculate it for him.

The caller can also provide a parentUncleHash. In most, but not all cases, the parent uncle hash also affects the formula. If no such hash is provided (or, if the empty all-zero hash is provided), it's assumed that there were no uncles.

holiman committed:
tests: avoid performance tests during CI runs (#23304)

fjl committed:
Dockerfile: upgrade to Go 1.17 (#23465)

gballet committed:
crypto/cloudflare/bn256: fix in-place addition and unmarshalling (#23419)

fjl committed:
all: add go:build lines (#23468)

Generated by go1.17 fmt ./...

fjl committed:
rlp: minor optimizations for slice/array encoding (#23467)

As per benchmark results below, these changes speed up encoding/decoding of
consensus objects a bit.

    name                             old time/op    new time/op    delta
    EncodeRLP/legacy-header-8           384ns ± 1%     331ns ± 3%  -13.83%  (p=0.000 n=7+8)
    EncodeRLP/london-header-8           411ns ± 1%     359ns ± 2%  -12.53%  (p=0.000 n=8+8)
    EncodeRLP/receipt-for-storage-8     251ns ± 0%     239ns ± 0%   -4.97%  (p=0.000 n=8+8)
    EncodeRLP/receipt-full-8            319ns ± 0%     300ns ± 0%   -5.89%  (p=0.000 n=8+7)
    EncodeRLP/legacy-transaction-8      389ns ± 1%     387ns ± 1%     ~     (p=0.099 n=8+8)
    EncodeRLP/access-transaction-8      607ns ± 0%     581ns ± 0%   -4.26%  (p=0.000 n=8+8)
    EncodeRLP/1559-transaction-8        627ns ± 0%     606ns ± 1%   -3.44%  (p=0.000 n=8+8)
    DecodeRLP/legacy-header-8           831ns ± 1%     813ns ± 1%   -2.20%  (p=0.000 n=8+8)
    DecodeRLP/london-header-8           824ns ± 0%     804ns ± 1%   -2.44%  (p=0.000 n=8+7)

* rlp: pass length to byteArrayBytes

This makes it possible to inline byteArrayBytes. For arrays, the length is known
at encoder construction time, so the call to v.Len() can be avoided.

* rlp: avoid IsNil for pointer encoding

It's actually cheaper to use Elem first, because it performs less checks
on the value. If the pointer was nil, the result of Elem is 'invalid'.

* rlp: minor optimizations for slice/array encoding

For empty slices/arrays, we can avoid storing a list header entry in the
encoder buffer. Also avoid doing the tail check at encoding time because
it is already known at encoder construction time.

baptiste-b-pegasys committed:
cmd/devp2p/internal/ethtest: remove WriteTo method (#23379)

WriteTo is unused, and go vet warns about it because it
doesn't match the usual signature of the io.WriterTo method.

zenovich committed:
rpc: add BlockNumber.MarshalText (#23324)

Currently rpc.BlockNumber is marshalled to JSON as a numeric value, which is
wrong because BlockNumber.UnmarshalJSON() wants it to either be hex-encoded
or string "earliest"/"latest"/"pending". As a result, the call chain

    rpc.BlockNumberOrHashWithNumber(123) -> json.Marshal() -> json.Unmarshal()

fails with error "cannot unmarshal object into Go value of type string".

fjl committed:
build: stop publishing deb packages for Ubuntu 20.10 Groovy Gorilla (#23470)

Support for this Ubuntu release has ended on July 22 2021 and
Launchpad no longer accepts uploads for it.

s1na committed:
node: remove dependency on wallet backend packages (#23019)

* accounts: new AddBackends method in manager

* node,cmd/geth: mv accman backend init to cmd/geth

* node,cmd/geth: mv scrypt config downstreawm from node

* accounts: use static buffer size for accman sub chan

minor fix

* accounts,cmd/geth: update accman backends through its event loop

* accounts,node: add comments

* accounts: un-export newBackendEvent

* accounts: use chan instead of wg in newBlockEvent

* node: rename isKeyDirEphem

* accounts,cmd: AddBackends->AddBackend

* accounts: fix potential blocking when adding backend

holiman committed:
Revert "eth, internal/ethapi: make RPC block miner field show block sealer correctly (#23312)" (#23466)

This reverts commit 57feabea663496109e59df669238398239438fb1.

lightclient committed:
internal/ethapi: support both input and data for personal_sendTransaction (#23476)

Currently, setDefaults overwrites the transaction input value if only input is provided. This causes personal_sendTransaction to estimate the gas based on a transaction with empty data. eth_estimateGas never calls setDefaults so it was unaffected by this.

Commit:
update chainId for tanenbaum

Commit:
Update config.go

Commit:
Update version.go

Commit:
Merge branch 'master' into nevm
